src/libraries/System.Text.Json/tests/Serialization/NumberHandlingTests.cs

```diff
@@ -0,0 +1,708 @@
+            ClassWithNumbers obj = JsonSerializer.Deserialize<ClassWithNumbers>(json, s_optionReadFromStrAllowFloatConstants);
+
+            Assert.Equal(1, obj.IntNumber);
+            Assert.Equal(float.NaN, obj.FloatNumber);
+
+            JsonTestHelper.AssertJsonEqual(@"{""IntNumber"":1,""FloatNumber"":""NaN""}", JsonSerializer.Serialize(obj, s_optionReadFromStrAllowFloatConstants));
+        }
+
+        [Fact]
+        public static void WriteAsString_AllowFloatingPoint()
+        {
+            string json = @"{""IntNumber"":""1"",""FloatNumber"":""NaN""}";
+            Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<ClassWithNumbers>(json, s_optionWriteAsStrAllowFloatConstants));
+
+            var obj = new ClassWithNumbers
+            {
+                IntNumber = 1,
+                FloatNumber = float.NaN
+            };
+
+            JsonTestHelper.AssertJsonEqual(json, JsonSerializer.Serialize(obj, s_optionWriteAsStrAllowFloatConstants));
+        }
+
+        public class ClassWithNumbers
+        {
+            public int IntNumber { get; set; }
+            public float FloatNumber { get; set; }
+        }
+
+        [Fact]
+        public static void FloatingPointConstants_IncompatibleNumber()
+        {
+            AssertFloatingPointIncompatible_Fails<byte>();
+            AssertFloatingPointIncompatible_Fails<sbyte>();
+            AssertFloatingPointIncompatible_Fails<short>();
+            AssertFloatingPointIncompatible_Fails<int>();
+            AssertFloatingPointIncompatible_Fails<long>();
+            AssertFloatingPointIncompatible_Fails<ushort>();
+            AssertFloatingPointIncompatible_Fails<uint>();
+            AssertFloatingPointIncompatible_Fails<ulong>();
+            AssertFloatingPointIncompatible_Fails<float>();
+            AssertFloatingPointIncompatible_Fails<decimal>();
+            AssertFloatingPointIncompatible_Fails<byte?>();
+            AssertFloatingPointIncompatible_Fails<sbyte?>();
+            AssertFloatingPointIncompatible_Fails<short?>();
+            AssertFloatingPointIncompatible_Fails<int?>();
+            AssertFloatingPointIncompatible_Fails<long?>();
+            AssertFloatingPointIncompatible_Fails<ushort?>();
+            AssertFloatingPointIncompatible_Fails<uint?>();
+            AssertFloatingPointIncompatible_Fails<ulong?>();
+            AssertFloatingPointIncompatible_Fails<float?>();
+            AssertFloatingPointIncompatible_Fails<decimal?>();
+        }
+
+        private static void AssertFloatingPointIncompatible_Fails<T>()
+        {
+            string[] testCases = new[]
+            {
+                "NaN",
+                "Infinity",
+                "-Infinity",
+            };
+
+            foreach (string test in testCases)
+            {
+                Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<T>(test, s_optionReadFromStrAllowFloatConstants));
```

Comment:
Doesn't this fails becuase the `Utf8JsonReader` failed to read the `test` string as JSON? Is that the intent here or are you missing wrapping `test` within quotes?

Reply:
I missed the quotes and need to remove the floating points from the calling test. This addresses https://github.com/dotnet/runtime/pull/39363/files#r455416625.